### PR TITLE
Completed Max_unpool1d implementation 

### PR DESCRIPTION
**Pull request by mohame54**
# PR Description 

**Completed max_unpool1d function implementation** I didn't add paddle backend implementation cuz paddle.max_unpool1d  had some errors related paddle itself

##

### CODE REVIEW

**Comment by ivy-leaves** [comment]
If you are working on an open task, please **edit** the PR description to link to the issue you've created.

For more information, please check [ToDo List Issues Guide](https://unify.ai/docs/ivy/overview/contributing/the_basics.html#todo-list-issues).

Thank you :hugs:

**Comment by mohame54** [comment]
my implementation is close to torch **2.0.1+cu118** version I don't know why gpu version of torch.max_unpool1d is different from the cpu version @AnnaTz

**Comment by AnnaTz** [comment]
> my implementation is close to torch **2.0.1+cu118** version I don't know why gpu version of torch.max_unpool1d is different from the cpu version @AnnaTz

I get this error when I run the test locally:
```
E       AssertionError:  the results from backend torch and ground truth framework jax do not match
E        [[[-1. -1. -1.  0.]
E         [-1. -1. -1.  0.]
E         [-1. -1. -1.  0.]
E         [-1. -1. -1.  0.]]
E       
E        [[-1. -1. -1.  0.]
E         [-1. -1. -1.  0.]
E         [-1. -1. -1.  0.]
E         [-1. -1. -1.  0.]]
E       
E        [[-1. -1. -1.  0.]
E         [-1. -1. -1.  0.]
E         [-1. -1. -1.  0.]
E         [-1.  0.  1.  0.]]]!=[[[-1. -1. -1.  0.]
E         [-1. -1. -1.  0.]
E         [-1. -1. -1.  0.]
E         [-1. -1. -1.  0.]]
E       
E        [[-1. -1. -1.  0.]
E         [-1. -1. -1.  0.]
E         [-1. -1. -1.  0.]
E         [-1. -1. -1.  0.]]
E       
E        [[-1. -1. -1.  0.]
E         [-1. -1. -1.  0.]
E         [-1. -1. -1.  0.]
E         [-1.  0.  2.  0.]]] 
E       
E       
E       Falsifying example: test_max_unpool1d(
E           backend_fw='torch',
E           on_device='cpu',
E           x_k_s_p=(['float32', 'int64'], array([[[-1., -1., -1.],
E                    [-1., -1., -1.],
E                    [-1., -1., -1.],
E                    [-1., -1., -1.]],
E            
E                   [[-1., -1., -1.],
E                    [-1., -1., -1.],
E                    [-1., -1., -1.],
E                    [-1., -1., -1.]],
E            
E                   [[-1., -1., -1.],
E                    [-1., -1., -1.],
E                    [-1., -1., -1.],
E                    [-1.,  1.,  2.]]], dtype=float32), array([[[0, 1, 2],
E                    [0, 1, 2],
E                    [0, 1, 2],
E                    [0, 1, 2]],
E            
E                   [[0, 1, 2],
E                    [0, 1, 2],
E                    [0, 1, 2],
E                    [0, 1, 2]],
E            
E                   [[0, 1, 2],
E                    [0, 1, 2],
E                    [0, 1, 2],
E                    [0, 2, 2]]]), (2,), (1,), 0),
```
It seems the failure happens when running on cpu, so I don't understand what you mean exactly.
Is there a hands-on example that shows the original torch function is inconsistent across versions?

**Comment by mohame54** [comment]
Yeah, I will show you one example.

**Comment by mohame54** [comment]
image

**Comment by mohame54** [comment]
you can see here that's torch **cuda** max_unpool  output from the same input above which is different  from the cpu version

**Comment by AnnaTz** [comment]
Ohh I get what's happening 😅
If you run the specific example many times it will sometimes return the correct value. The result is non-deterministic because the indices contain repeated values.
They actually mention it in their documentation 
image

**Comment by mohame54** [comment]
Yeah, that's a catch I tried before to remove the repeated indices but it was not useful also the tests were giving me errors

**Comment by AnnaTz** [comment]
> Yeah, that's a catch I tried before to remove the repeated indices but it was not useful also the tests were giving me errors

What kind of errors? I think you should be able to force the indices to not include identical values. We should at least know if that's the only problem (if the test passes for all other cases) before we decide what to do about it.

**Comment by mohame54** [comment]
output mismatch errors if tried to exclude the repeated indices there are mismatch errors with torch backend as you can see so I didn't add  the excluding part to the implementation .

**Comment by mohame54** [comment]
>What kind of errors? I think you should be able to force the indices to not include identical values. We should at least know if that's the only problem (if the test passes for all other cases) before we decide what to do about it.
I tried as many tests as I could but the only error that appeared to me were output mismatching errors.

**Comment by mohame54** [comment]
Also I didn't complete the paddle backend implementation cuz the paddle function was giving me errors also  and when I tried it on colab with  `paddle 2.4.0` installed it didn't work for me. it seems that the  function has not been implemented yet.

**Comment by AnnaTz** [comment]
I'm inclined to say we will need a compositional implementation for this 💀

**Comment by mohame54** [comment]
->I'm inclined to say we will need a compositional implementation for this 💀
but Why?

**Comment by mohame54** [comment]
Do you mean compositional implementation using ivy API instead of backend specific API wouldn't this approach affect the performance 😕

**Comment by AnnaTz** [comment]
> ->I'm inclined to say we will need a compositional implementation for this 💀 but Why?

We need an implementation that works consistently for all backends in all cases. The native torch function does not work as expected in the repeated indices case, and the paddle function is not implemented in some versions and giving errors in others. So We need an ivy function implementation to cover these.

**Comment by AnnaTz** [comment]
> Do you mean compositional implementation using ivy API instead of backend specific API, wouldn't that affect the performance 😕

Also, about that, according to what @vedpatwardhan has told me, I think the implementation we have for the numpy, jax and tensorflow backends could be replaced by an ivy composition. It seems that we are doing the exact same operations in all these backends, so it would be better to avoid code repetition.

**Comment by mohame54** [comment]
But I agree with you in order to reduce the code repetition  we need compositional implementation

**Comment by AnnaTz** [comment]
> But I agree with you in order to reduce the code repetition we need compositional implementation

I think it will be quite simple since you have already implemented the backend functions compositionally. I think you will mainly copy that and replace native calls with ivy calls.

**Comment by mohame54** [comment]
sorry I didn't mean to edit your comment so sorry @AnnaTz 😄

File: ivy/data_classes/array/experimental/layers.py [hunk 1]
```diff
@@ -1117,3 +1117,45 @@ def rfftn(
             The result of the RFFT operation.
         """
         return ivy.rfftn(self._data, s=s, axes=axes, norm=norm, out=out)
+
+    def max_unpool1d(
+        self: ivy.Array,
+        indices: ivy.Array,
+        kernel_size: Union[Tuple[int], int],
+        strides: Union[int, Tuple[int]] = None,
```
Comment (AnnaTz) [inline ivy/data_classes/array/experimental/layers.py:1125]: Just noticed you need to add `/, *,` before the start of the keyword default arguments.

File: ivy/data_classes/array/experimental/layers.py [hunk 2]
```diff
@@ -1117,3 +1117,45 @@ def rfftn(
...
+        padding: Union[int, Tuple[int]] = 0,
```
Comment (AnnaTz) [inline ivy/data_classes/array/experimental/layers.py:1126]: Also, please remove the `docs/demos` changes.

File: ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_layers.py [hunk 3]
```diff
@@ -1190,23 +1187,54 @@ def test_max_pool3d(
...
+            data_format="channel_first",
+            **data_gen_kwargs)
+    )
+    dts.extend(["int64"])
+    values = values[0]
+    padding = draw(helpers.ints(min_value=0, max_value=2))
+    if padding > (kernel_size[0] // 2):
+        padding = 0
+
+    values, indices = torch.nn.functional.max_pool1d(
```
Comment (AnnaTz) [inline ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_layers.py:1208]: @sherry30 are we allowed to use native functions in testing?

File: ivy/functional/ivy/experimental/layers.py [hunk 4]
```diff
@@ -180,9 +205,45 @@ def max_unpool1d(
...
+    f = one_like_mask * feature_range
+    updates_size = input.size
+    indices = ivy.stack([b, f, indices]).reshape(3, updates_size)
+    output = ivy.zeros(output_shape, dtype=input.dtype)
+    indices = tuple(indices)
+    values = input.reshape((-1,))
+    output[indices] = values
+    if revert:
+        output = output.permute_dims([0, 2, 1])
+    return output
```
Comment (AnnaTz) [inline ivy/functional/ivy/experimental/layers.py:247]: The code looks good overall. The only thing remaining is to add the partial torch backend.

**Review by AnnaTz — CHANGES_REQUESTED** [review]

File: ivy/functional/ivy/experimental/layers.py (hunk 4 above)
Comment (mohame54) [reply ivy/functional/ivy/experimental/layers.py:247]: What do you mean by partial torch backend? I did not get what you meant here

File: ivy/data_classes/array/experimental/layers.py (hunk 2 above)
Comment (mohame54) [reply ivy/data_classes/array/experimental/layers.py:1126]: If I remove them there will be conflicts with the main branch

File: ivy/functional/ivy/experimental/layers.py (hunk 4 above)
Comment (AnnaTz) [reply ivy/functional/ivy/experimental/layers.py:247]: https://unify.ai/docs/ivy/overview/deep_dive/function_types.html#partial-mixed-functions

File: ivy/data_classes/container/experimental/layers.py [hunk 5]
```diff
@@ -2280,3 +2280,112 @@ def rfftn(
...
+    @staticmethod
+    def static_max_unpool1d(
+        input: ivy.Container,
+        indices: ivy.Container,
+        kernel_size: Union[Tuple[int], int],
+        strides: Union[int, Tuple[int]] = None,
+        padding: Union[int, Tuple[int]] = 0,
+        data_format: Union[str, ivy.Container] = "NWC",
+        /,
+        *,
```
Comment (AnnaTz) [inline ivy/data_classes/container/experimental/layers.py:2293]: Misplaced `/, *` here (sorry to nitpick 😅)

**Review by AnnaTz — CHANGES_REQUESTED** [review]

File: ivy/data_classes/container/experimental/layers.py (hunk 5 above)
Comment (mohame54) [reply ivy/data_classes/container/experimental/layers.py:2293]: okay no problem.

**Pull request closed without merging**